pkg/lib/cockpit-components-dropdown.tsx

```diff
@@ -35,49 +34,110 @@ import { EllipsisVIcon } from '@patternfly/react-icons';
+        const pathSansLast = drilldownItemPath.slice(0, drilldownItemPath.length - 1);
+        setDrilledInMenus(drilledInMenusSansLast);
+        setDrilldownItemPath(pathSansLast);
+        setActiveMenu(toId);
+    }
+
+    function resetDrilldown() {
+        setDrilledInMenus([]);
+        setDrilldownItemPath([]);
+        setActiveMenu(id || "");
+    }
+
+    function onGetMenuHeight(menuId: string, height: number) {
+        if (menuHeights[menuId] === undefined || (menuId !== id && menuHeights[menuId] !== height)) {
+            setMenuHeights({ ...menuHeights, [menuId]: height });
+        }
+    }
+
+    function onSelect(_event: React.MouseEvent<Element, MouseEvent> | undefined, itemId: string | number | undefined) {
+        if (!itemId || typeof itemId == "number" || !itemId.startsWith("drilldown:")) {
+            setKebabOpen(false);
+            resetDrilldown();
+        }
+    }
+
     return (
         <Dropdown
+            {... id ? { id } : { } }
             onOpenChange={isOpen => setKebabOpen(isOpen)}
-            onSelect={() => setKebabOpen(false)}
+            onSelect={onSelect}
             toggle={(toggleRef) => (
                 <MenuToggle
                     id={toggleButtonId}
                     isDisabled={isDisabled}
                     ref={toggleRef}
                     variant="plain"
-                    onClick={() => setKebabOpen(!isKebabOpen)}
+                    onClick={() => {
+                        setKebabOpen(!isKebabOpen);
+                        resetDrilldown();
+                    }}
                     isExpanded={isKebabOpen}
                 >
                     <EllipsisVIcon />
                 </MenuToggle>
             )}
             isOpen={isKebabOpen}
-            popperProps={{ position }}
+            popperProps={{ position, ...popperProps }}
+            {... (containsDrilldown
+                ? {
+                    containsDrilldown: true,
+                    drilldownItemPath,
+                    drilledInMenus,
+                    onDrillIn,
+                    onDrillOut,
+                    activeMenu,
+                    onGetMenuHeight,
+                    menuHeight: `${menuHeights[activeMenu]}px`,
+                    // Setting menuHeight forces isScrollable to be true, but we don't
+                    // want that so we compensate by setting maxMenuHeight to "infinitiy".
+                    maxMenuHeight: "10000px",
+                }
+                : { })
+            }
         >
             <DropdownList>
                 {dropdownItems}
             </DropdownList>
         </Dropdown>
     );
 };
-
-KebabDropdown.propTypes = {
-    dropdownItems: PropTypes.array.isRequired,
-    isDisabled: PropTypes.bool,
-    toggleButtonId: PropTypes.string,
-    position: PropTypes.oneOf(['right', 'left', 'center', 'start', 'end']),
-    isOpen: PropTypes.bool,
-    setIsOpen: PropTypes.func,
-};
```

Comment:
We still have non typescript projects around, so I don't understand why we remove this.

Reply:
Do we actually enforce these annotations? I have found so many wrong ones that I thought they are all rotting away slowly. But I am happy to keep this.

Reply:
At least the intention is "yes". The /devel scenario enables React runtime errors/warnings and our tests fail on that. testlib.py `default_allowed_console_errors` has a few React errors that we ignore (we never got around to fixing the rest of it), but wrong prop types should cause an error. Of course that may just have bitrotted.